[Hub Generated] Review request for Microsoft.Web to add version 2018-02-01

dejo-msft opened the pull request:
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/app/branch/dejo-msft/azure-rest-api-specs/dev-web-Microsoft.Web-2018-02-01).

### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

dejo-msft committed:
Add Hosting Environment network dependencies

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-python

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-java

AutorestCI commented:
# Automation for azure-sdk-for-node

praries880 commented:
@dejo-msft Can you add examples for the new API you added in the PR?

antmarti-microsoft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -2177,6 +2306,30 @@
+      "type": "object",
+      "properties": {
+        "description": {
+          "description": "Text describing the endpoints.",
```
Please add a description

dejo-msft replied:
This is a short description of what the endpoints is accessed for. E.G. "App Service management", "App Service Environment VIP", or "App Service Environment subnet".
Does this more accurately represent this? 
"Short description of why the endpoints are accessed."

antmarti-microsoft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -2177,6 +2306,30 @@
+            "type": "string"
+          }
+        },
+        "ports": {
```
Given that endpoints and ports are separate arrays, how does this work in practice? Each endpoint IP address supports each port?

antmarti-microsoft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -2163,6 +2253,45 @@
+          "type": "number"
+        },
+        "isAccessable": {
+          "description": "Whether it is possible to connect to IpAddress.",
```
What does this mean? Whether you are able to make an inbound TCP connection to the ipAddress + port? From where?

antmarti-microsoft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -2231,6 +2384,23 @@
+      "type": "object",
+      "properties": {
+        "category": {
+          "description": "Short description of the endpoints.",
```
Fill out description

antmarti-microsoft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -2231,6 +2384,23 @@
       }
     },
+    "OutboundEnvironmentEndpoint": {
+      "description": "Endpoints of a common type.",
```
What does this mean? Please can you be more specific with the description?

antmarti-microsoft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -2231,6 +2384,23 @@
         }
       }
     },
+    "OutboundEnvironmentEndpoint": {
```
This is a similar model to "InboundEnvironmentEndpoint" except it doesn't include the ports. Is this intentional?

dejo-msft replied:
Yes, this is intentional. Inbound endpoints intentionally leave out the Fully Qualified Domain Name to avoid including infrastructure implementation details. Outbound endpoints include the FQDN so that they can be whitelisted by a layer 7 firewall. We don't have the same support for inbound endpoints so we only have the IP Address and port.

antmarti-microsoft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -539,6 +539,51 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/hostingEnvironments/{name}/inboundnetworkdependenciesendpoints": {
```
inboundnetworkdependenciesendpoints feels quite long-winded and it's not obvious to me what this actually is. Let's have a discussion to go over some of the naming.

dejo-msft replied:
Unfortunately, as I mentioned before, this API is in production and the signature can't be changed in this version. I'll work on options in the API name.  Does dropping "Endpoints" make more sense? Initially, the api was a list of <IP, Port> but the requirements changed. InboundNetworkDependencies and OutboundNetworkDependencies?

KrisBash replied:
Understood. Would you please lowerCamelCase the type name? `inboundNetworkDependenciesEndpoints`

praries880 commented:
@dejo-msft Pinging back... are you working on the ARM feedback

ravbhatnagar commented:
@dejo-msft - please let us know when its ready for ARM re-review.

dejo-msft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -2163,6 +2253,45 @@
+          "description": "The time in milliseconds it takes to connect to this IpAddress at this Port.",
+          "type": "number"
+        },
+        "isAccessable": {
```
Unfortunately, this API is in production and in use by customers and the Azure Portal. I've created an issue here https://github.com/Azure/azure-rest-api-specs/issues/5255 which will require a new API version to fix.

derekj627 committed:
Address PR description feedback

dejo-msft committed:
Add Network Dependencies examples

dejo-msft commented:
I've addressed feedback on the description texts and added examples of the API. 
@ravbhatnagar and @praries880 This is ready for ARM review again. Sorry for the delay.

praries880 commented:
@dejo-msft  kindly fix the [linter errors](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/499038791#L815)

KrisBash commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -1209,6 +1259,56 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/hostingEnvironments/{name}/outboundnetworkdependenciesendpoints": {
```
While not functionally impactful given case insensitivity in URIs, the resourceTypeNames should follow lowerCamelCase convention

dejo-msft replied:
Fixed these. Thanks!

KrisBash commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -2163,6 +2263,47 @@
+      "type": "object",
+      "properties": {
+        "domainName": {
+          "description": "The Domain Name of the dependency.",
```
Nit - inconsistent casing of "Domain Name" in the EndpointDependency description vs domainName property description

KrisBash reviewed (approved):
Feedback looks to be addressed. i've added a few non-blocking comments regarding casing..

dejo-msft commented:
@sergey-shandar you pushed a fix on https://github.com/Azure/azure-rest-api-specs/pull/5288 to the linterdiff tool. Can you do the same here?

praries880 commented:
@dejo-msft  kindly fix the CI issues, we have ARM signoff... We wont be able to merge this in till the CI is fixed.

dejo-msft committed:
Fix casing

praries880 commented:
@antmarti-microsoft  can you kindly sign off the PR, i cant merge this until you sign off :)

antmarti-microsoft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -539,6 +539,56 @@
+        },
+        "x-ms-examples": {
+          "Get Inbound Network Dependencies Endpoints": {
+            "$ref": "./examples/GetInboundNetworkDependenciesEndpoints.json"
```
The example uses paging, but this has not been called out here with an "x-ms-pageable" declaration. Please can you either update the example, or the schema, depending on whether or not you are using paging on this response?

dejo-msft replied:
The APIs are not paged. I've removed them from examples. Thanks for catching this!

antmarti-microsoft commented on specification/web/resource-manager/Microsoft.Web/stable/2018-02-01/AppServiceEnvironments.json:
```diff
@@ -1209,6 +1259,56 @@
+        },
+        "x-ms-examples": {
+          "Get Outbound Network Dependencies Endpoints": {
+            "$ref": "./examples/GetOutboundNetworkDependenciesEndpoints.json"
```
The example uses paging, but this has not been called out here with an "x-ms-pageable" declaration. Please can you either update the example, or the schema, depending on whether or not you are using paging on this response?

anthony-c-martin reviewed (changes requested):
Please address the discrepancy between examples (using paging), and the response declaration (not using paging).

dejo-msft committed:
Remove incorrect pagination from examples.
